Add default_options argument to find_program()

Nomura-RH opened the pull request:
This PR adds a `default_options` argument to `find_program()`. It allows users to pass options to the subproject in case of a fallback , in a similar fashion as the argument of the same name in `dependency()`. If this feature is acceptable, I will proceed to write a test for it.

xclaesse commented on mesonbuild/interpreter/interpreter.py:
```diff
@@ -1720,6 +1723,7 @@ def find_program_fallback(self, fallback: str, args: T.List[mesonlib.FileOrStrin
         REQUIRED_KW,
         KwargInfo('dirs', ContainerTypeInfo(list, str), default=[], listify=True, since='0.53.0'),
         KwargInfo('version', ContainerTypeInfo(list, str), default=[], listify=True, since='0.52.0'),
+        DEFAULT_OPTIONS.evolve(since='1.3.0'), allow_unknown=True
```
Why adding `allow_unknown=True`?

Nomura-RH replied:
I copied the statement from the kwargs of the dependency() function

eli-schwartz replied:
Well, those are two different arguments, only one of which is relevant here... :)

Nomura-RH replied:
Didn't realize they were unrelated. I removed allow_unknown :)

xclaesse commented:
Makes sense, yes. Missing:
- test
- doc
- release notes

But implementation looks good.

Nomura-RH commented:
Added release note, updated `find_program`'s YAML doc and made minor changes to make pylint happy. 

I'm unsure what the test should look like, considering `find_program()` only falls back to wrap subprojects. Should I use a wrap from WrapDB or my own?

Nomura-RH commented:
Nevermind, I was able to come up with a test. Anything else that needs to be done before this is merged?

xclaesse commented:
Looks all good to me, let's see if CI pass.

Nomura-RH commented:
> Looks all good to me, let's see if CI pass.

Looks like run_project_tests.py doesn't like my dummy program. I'll add 'void' to main(), let's see if that does the trick

tristan957 commented:
@xclaesse looks like we got the updates you requested

xclaesse commented on mesonbuild/interpreter/interpreter.py:
```diff
@@ -1617,6 +1617,7 @@ def notfound_program(self, args: T.List[mesonlib.FileOrString]) -> ExternalProgr
     # the host machine.
     def find_program_impl(self, args: T.List[mesonlib.FileOrString],
                           for_machine: MachineChoice = MachineChoice.HOST,
+                          default_options: T.Dict[OptionKey, T.Union[str, int, bool, T.List[str]]] | None = None,
```
We usually use T.Union[foo, bar] instead of `foo | bar`, but I don't know why. Maybe for compat with older python version? @dcbaker @eli-schwartz ^

xclaesse replied:
In this case it's rather `T.Optional[T.Dict[...]]`

tristan957 replied:
Yes, that is the case. https://peps.python.org/pep-0604/

Nomura-RH replied:
Fixed
